### DISCUSSION TITLE
Some formula fail to install or upgrade on M1 MacBook Pro Ventura 13.4 (unable to run /usr/local/sbin/pw: No such file or directory)

### DISCUSSION THREAD

**Discussion by theclive** [Homebrew/discussions#4617] [Everyday usage] [answered]
### Output of `brew config`

```shell
➜ brew config
HOMEBREW_VERSION: 4.0.26
ORIGIN: https://github.com/Homebrew/brew
HEAD: eff45ef570f265e226f14ce91da72d7a6e7d516a
Last commit: 6 days ago
Core tap JSON: 03 Jul 09:34 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
SUDO_ASKPASS: /usr/local/sbin/pw
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
➜ brew doctor
Your system is ready to brew.
```

### Description of issue

I get this same error, `unable to run /usr/local/sbin/pw: No such file or directory`, whenever brew needs to escalate privileges (sudo). I've only recently received this brand new MacBook (within the last three weeks) and installed brew as described on https://brew.sh/

I have 19 casks and 32 normal formula installed.  I was unable to install some formula, and initially I just assumed the formula was broken.  Now, however, when I am doing my first `brew upgrade` and receiving the error for formula that are already installed, I suspect an issue with my installation/config.  I already followed [these initial steps](https://docs.brew.sh/Troubleshooting), and now I am here, posting on the discussion forum.

Here is the output from `brew upgrade`:
```
➜ brew upgrade
==> Casks with 'auto_updates true' or 'version :latest' will not be upgraded; pass `--greedy` to upgrade them.
==> Upgrading 1 outdated package:
google-cloud-sdk 434.0.0 -> 437.0.1
==> Upgrading google-cloud-sdk
==> Caveats
To add gcloud components to your PATH, add this to your profile:

  for bash users
    source "$(brew --prefix)/share/google-cloud-sdk/path.bash.inc"

  for zsh users
    source "$(brew --prefix)/share/google-cloud-sdk/path.zsh.inc"
    source "$(brew --prefix)/share/google-cloud-sdk/completion.zsh.inc"

  for fish users
    source "$(brew --prefix)/share/google-cloud-sdk/path.fish.inc"

==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/8bd6c63b335d7fa5a77f0d82a90c73a4eb4afd57/Casks/google-cloud-sdk.rb
Already downloaded: /Users/clive/Library/Caches/Homebrew/downloads/9d95ef5ea452331ef135292074bacb52c095a6a8ab1025398efaf648040a5e9f--google-cloud-sdk.rb
==> Downloading https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-cli-437.0.1-darwin-arm.tar.gz
Already downloaded: /Users/clive/Library/Caches/Homebrew/downloads/4714a15f885bdd366dd7f79578d84e4a65bca2246aac9543fc9c3f55d06b5fa4--google-cloud-cli-437.0.1-darwin-arm.tar.gz
All formula dependencies satisfied.
==> Removing files:
/opt/homebrew/Caskroom/google-cloud-sdk/latest
sudo: unable to run /usr/local/sbin/pw: No such file or directory
sudo: no password was provided
sudo: a password is required
==> Purging files for version 437.0.1 of Cask google-cloud-sdk
Error: google-cloud-sdk: Failure while executing; `/usr/bin/sudo -A -E -- /usr/bin/xargs -0 -- /bin/rm -r -f --` exited with 1. Here's the output:
sudo: unable to run /usr/local/sbin/pw: No such file or directory
sudo: no password was provided
sudo: a password is required
```

That `pw` file does not exist:
```
➜ ls -lh /usr/local/sbin/pw
ls: /usr/local/sbin/pw: No such file or directory
```

And a binary called `pw` does not seem to be installed:
```
➜ which pw
pw not found
```

Have I somehow missed a dependency for installing brew? 
Pls help.

**Comment by SMillerDev** [comment]
`sudo: unable to run /usr/local/sbin/pw: No such file or directory` this means your sudo configuration defines the usage of `/usr/local/sbin/pw` which has nothing to do with brew

**Comment by theclive** [comment, marked as answer]
I've solved the problem.  I decided to do a search on `SUDO_ASKPASS` (which I saw in `brew config`) and found [this link](https://stackoverflow.com/questions/12608293/how-to-setup-a-sudo-askpass-environment-variable).

In summary, `SUDO_ASKPASS` (as shown in the `brew config`) must contain a path to an executable that outputs the password to stdout that you want to use for `sudo` privilege escalation.  So all I needed to do was to create a bash script at `/usr/local/sbin/pw` that would output my Mac password.

To avoid saving my password in plain text in the bash script, since I use 1Password, I made use of the [1Password CLI](https://developer.1password.com/docs/cli/) to fetch the PWD.  It prompts for my biometrics so I have the option to deny if it is not something I prompted my computer to do.

One thing that I was still puzzled by was, why was `SUDO_ASKPASS` set on a new mac that I had just received.  Well, this is where I was being an idiot.  When I got my new MacBook, one of the first things i did was copied over my old `~/.zshrc` file from my old MacBook, and it was defined in there. 🙈   I must have forgotten about it from some time ago.
(This explains why I couldn't find anyone else with this problem. That path was unique to me 🙄 🤣 ).

Nonetheless, I hope the existence of this post will help someone in the future.
